### PR TITLE
📥 Make Total Spaces Really Cool 😎

### LINKED ISSUE #348: 📥 Some thoughts about making it cooler to work with the total space of type family
## Context
The main motivation (as I understand it) for introducing the ability to treat a type family as its total space was to bundle up parameters/indices that would normally have to be quantified over, making it easier to work with heavily parametrized/indexed types without having to implement implicit arguments and unification. cooltt is now at a point where the _type_ of a structure can be conveniently written down using total space conversion and record patching!
```
def category : type :=
  sig (ob : type)
      (hom : sig (s : ob) (t : ob) -> type)
      (idn : (x : ob) -> hom # [ s .= x | t .= x ])
      (seq : (f : hom) -> (g : hom # [ s .= f.t ]) -> hom # [ s .= f.s | t .= g.t ])
      (seqL : (f : hom) -> path {hom # [ s .= f.s | t .= f.t ]} {seq {idn {f.s}} f} f)
      (seqR : (f : hom) -> path {hom # [ s .= f.s | t .= f.t ]} {seq f {idn {f.t}}} f)
      (seqA : (f : hom) -> (g : hom # [ s .= f.t ]) -> (h : hom # [ s .= g.t ]) -> path {hom # [ s .= f.s | t .= h.t ]} {seq f {seq g h}} {seq {seq f g} h})
```
This is (to me, at least) just as easy to write and read as the equivalent definition using implicit arguments. It might even be easier, actually. Annoyingly, actually inhabiting this type is _not_ as nice as inhabiting the version using implicit arguments. Currently the definition of the category of cooltt types looks like this:
```
def types : category :=
  struct (ob : type)
         (hom : args => {args.s} -> {args.t})
         (idn : x => struct (fib : x => x))
         (seq : f g => struct (fib : x => g.fib {f.fib x}))
         (seqL : f i => f)
         (seqR : f i => f)
         (seqA : f g h i => struct (fib : x => {h.fib} {{g.fib} {{f.fib} x}}))
```
We're forced to construct these one element structs and project out their `fib` when we want to use them as actual `hom`s and not as the total space of a `hom`. Ideally, it would be possible to write this:
```
def types : category :=
  struct (ob : type)
         (hom : args => {args.s} -> {args.t})
         (idn : x z => z)
         (seq : f g x => g {f x}))
         (seqL : f i => f)
         (seqR : f i => f)
         (seqA : f g h i x => h {g {f x}}))
```
And have all the `struct (fib : ...)` and `... .fib` automatically inserting during elaboration. I've been thinking about how to do this, and unfortunately it is not as straight forward as I had hoped. Here's what I've come up with so far.

## Thoughts about making total spaces cooler

* If we are checking against a total space where every field except `fib` has been patched, then inserting `struct (fib: ...)` and filling in the rest of fields is easy. This can be straightforwardly added to the `intro_implicit_connectives` tactic. 

* If we are checking against a total space that does not have every field patched, we're basically out of luck, and must write down a record. This makes sense if the `fib` field is some term that must be checked, like a lambda, but if the `fib` field is synthable, it _feels_ like we ought to be able to propagate that information out of the term to fill in the unpatched fields, but I do not see a way to actually do this. If the `fib` field synths to `fam {struct ...}`, then we could use that. But if the `fam` in question is not neutral, we pretty much end up back to needing unification to figure out the values of the unpatched fields.

* Automatically inserting `.fib` projections is slightly problematic. My feeling is that we'd want to insert them as part of an `elim_implicit_connectives` style tactic, but _not_ if we're synthing a term to be projected. The reason for this is that we need to be able to project out the other fields of total space (`s` and `t` in the case of `hom`). So given `f : hom`: `f x` becomes `{f.fib} x` but `f.s` does _not_ become `{f.fib}.s`. This means that if the value of the `fib` field is a record, we would be required to manually insert the `.fib` projection. 

* The way that checking and synthing interact is problematic. If we have `f : hom # [s .= s | t .= t]`, and we go to check it against `hom # [s .= s | t .= t]`, then all is all well. The goal becomes `hom {struct (s : s) (t : t)` and `f` becomes `f.fib`. We run the `intro_fields` tactics with `struct (fib : f.fib)` and the other fields are filled in. But (!) if we have `f : hom # [s .= s | t .= t]` and we go to check it against `hom` (no patches), we're in trouble. `f` becomes `f.fib`, but as previously mentioned, we can't implicitly insert a struct when checking against a total space that isn't completely patched, so we get a type error. But this would have worked if we had not inserted the `.fib`. So we'd need to only insert it if we know that we aren't ultimately going to check against a total space.

I realize these thoughts are not really fully formed. If this is too rambly to be useful, let me know. I would love some input on whether these problems are fundamental, or I'm just approaching this in the wrong way.

### PR DESCRIPTION
**Pull request by mmcqd**
### New feature
This PR implements the proposal roughly described in #348 to make total spaces totally (lol) implicit whenever possible. We can now write down the definition of the category of types like this:
```
def types : category :=
  struct (ob : type)
         (hom : args => {args.s} -> {args.t})
         (idn : x z => z)
         (seq : f g x => g {f x})
         (seqL : f i => f)
         (seqR : f i => f)
         (seqA : f g h i x => h {g {f x}})
```
(Closes #348)
### How this is done
* Total spaces where all fields but `fib` have been patched are handled by `intro_implicit_connectives`. The user need only write down the value of the `fib` field
* Total spaces with unpatched fields are still just regular signature types, and the user must write down a record to inhabit them
* Any term that synths to a total space will have it's `fib` field implicitly projected, _unless_ the term was going to be projected from by the user. That is, `e .lbl` will not implicitly become `{e .fib} .lbl` if `e` synths to a total space. This allows us to project out the bundled fields of a total space. The one negative consequence of this is that if your `fib` field is itself a record, you must manually insert `.fib` projections. Perhaps we could introduce some special syntax to project total space fields?

The concern about the interaction of checking and synthing I mentioned in #348 happily turned out to be a non-issue. My experience was that it just magically worked haha. I'm pretty sure it's the eta-expansion during elaboration that handles it.

### Remaining issues
* **DONE:** This PR needs more tests
* I feel that the `elim_total` optional argument I've added to `syn_tm` is a little awkward/inelegant, thought it does work. Perhaps it should be part of the local state?
* **DONE:** When displaying a hole where the goal is a fully patched total space, the user should see the type of the `fib` field as the goal, but currently the whole signature is shown

### CODE REVIEW

**Comment by favonia** [comment]
:genie: question: Why can't `(hom : args => {args.s} -> {args.t})` be `(hom : args => args.s -> args.t)`?

**Comment by mmcqd** [comment]
> 🧞 question: Why can't `(hom : args => {args.s} -> {args.t})` be `(hom : args => args.s -> args.t)`?

I wrote it like that because that's how the test was previously. Upon trying, it looks like `args.s -> args.t` is parsed as `{args.s -> args}.t`, so `args.s -> {args.t}` works, but your suggestion does not. It would be nice if that worked though! Definitely a task for the 🧞

File: src/frontend/Tactics.mli [hunk 1]
```diff
@@ -10,9 +10,12 @@ module R := Refiner
 
 open Tactic
 
+val is_total : D.sign -> [`TotalAll | `TotalSome | `NotTotal] RM.m
```
Comment (TOTBWF) [inline src/frontend/Tactics.mli:13]: A short comment here describing the meaning of each of the return types would be handy :)

File: src/frontend/Tactics.ml [hunk 2]
```diff
@@ -13,6 +13,27 @@ module TB = TermBuilder
 
 open Monad.Notation (RM)
 
+
+let is_total (sign : D.sign) = 
+  let rec go acc = function
+    | D.Field (`User ["fib"],_,D.Clo ([],_)) -> RM.ret @@ acc
```
Comment (TOTBWF) [inline src/frontend/Tactics.ml:19]: We should probably change the magic `fib` field to be `Machine` at some point. Out of scope for this PR though, so don't worry about it :)

**Review by TOTBWF — APPROVED** [review]
Looks great! Only have a single nit regarding some documentation, but this is some really fantastic work!

**Comment by favonia** [comment]
@mmcqd I reformatted your code.

**Comment by mmcqd** [comment]
@favonia ack, thank you! I meant to do this before it got merged, sorry for not being consistent with using ocp-indent on my PRs.

File: src/frontend/Tactics.mli (hunk 1 above)
Comment (mmcqd) [reply src/frontend/Tactics.mli:18]: done

**Comment by mmcqd** [comment]
I've added more tests and support for auto-inserting fully patched total spaces at holes

**Pull request merged**